src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -184,9 +184,12 @@ public override Iterator<TSource> Clone()
 
             public override void Dispose()
             {
-                if (_enumerator is IDisposable) ((IDisposable)_enumerator).Dispose();
-                _enumerator = null;
-                base.Dispose();
+                if (_enumerator != null)
+                {
+                    ((IDisposable)_enumerator).Dispose();
+                    _enumerator = null;
+                    base.Dispose();
```

Comment:
I'm still unclear what case you're trying to protect against by explicitly casting the ```IEnumerator<T>``` to ```IDisposable<T>```.  Can you provide an example?  This change is also no longer calling Dispose on the base in the case where ```MoveNext()``` was never called to initialize ```_enumerator```.